common/proto/dex_amm.proto

```diff
@@ -6,22 +6,135 @@ import "erc20.proto";
 package messari.dex_amm.v1;
 
 message Pool {
-  string name = 1;
-  string address = 2; // bytes: Address
-  repeated messari.erc20.v1.ERC20Token input_tokens = 3;
-
-  // Metrics
-  string total_value_locked = 100; // string: BigDecimal
-}
-
-message PoolToken {
   string address = 1; // bytes: Address
-  // Balance of input token in native amounts
-  string balance = 2; // string: BigInt
-  // Weights of input token in the liquidity pool in percentage values. For example, 0.5/0.5 for Uniswap pools, 0.482/0.518 for a Curve pool, 0.1/0.1/0.8 for a Balancer pool
-  double weight = 3;
+  string name = 3;
+  uint64 created_timestamp = 5; // uint64: Timestamp
+  uint64 created_block_number = 6; // uint64: Block
+  repeated string input_tokens = 7;
+  repeated string input_token_amounts = 9;
+  string output_token = 10;
+  string output_token_supply = 11;
+  string total_value_locked = 12;
 }
 
 message Pools {
   repeated Pool items = 1;
 }
+
+message Deposit {
+  // Transaction hash of the transaction that emitted this event
+  string tx_hash = 1;
+
+  // Event log index. For transactions that don't emit event, create arbitrary index starting from 0
+  uint32 log_index = 2;
+
+  // Address that received the tokens
+  string to = 3;
+
+  // Address that sent the tokens
+  string from = 4;
+
+  // Block number of this event
+  uint64 block_number = 5;
+
+  // Timestamp of this event
+  uint64 timestamp = 6;
+
+  // Input tokens of the pool. E.g. WETH and USDC to a WETH-USDC pool
+  repeated string inputTokens = 7;
+
+  // Output token of the pool. E.g. the UNI-LP token
+  string outputToken = 8;
+
+  // Amount of input tokens in the token's native unit
+  repeated string input_token_amounts = 9;
+
+  // Amount of output tokens in the token's native unit
+  string output_token_amount = 10;
+
+  // The pool involving this transaction
+  string pool_address = 11;
+}
+
+message Deposits {
+  repeated Deposit items = 1;
+}
+
+message Withdraw {
+  // Transaction hash of the transaction that emitted this event
+  string tx_hash = 1;
```

Comment:
I've been thinking about creating a container `Transaction` struct that would include fields like `tx_hash`, `block_number`, and `timestamp` so we don't need to repeat them for each message. What are your thoughts?

Reply:
Yeah! I hadn't had time to push the changes yet, but I have done something like this. I just chose the name `UsageEvents`, but it can be changed to whatever seems most appropriate. I did not use transaction, because multiple of these events can occur with a single transaction.

Reply:
Actually, using `Transaction` as an abstraction above `UsageEvents` might be a good idea. Then we can more easily keep track of transaction counts in the `UsageMetricsDailySnapshots` as well.

Reply:
It would be cleaner to build this data structure if we mapped all the raw event data together in a single mapping module instead of `map_swap_events`, `map_deposit_events`, etc, because you would then have to group the events back together by transaction hash. But it trades off with not being able to immediately work with specific raw event data in isolation, since you will have to first reprocess the `Transaction` message to get back out the event data that you want to work with.

Reply:
Here is an option mapping raw events to the `Transaction` and `Event` abstractions. 
<img width="746" alt="Screen Shot 2022-11-06 at 8 23 44 PM" src="https://user-images.githubusercontent.com/56660047/200213644-ef3aad2c-edab-4c8b-8038-5d58ad463823.png">

Or an option using a similar scheme, but for the final `Deposit`, `Withdraw`, and `Swap` messages.
<img width="740" alt="Screen Shot 2022-11-06 at 8 30 13 PM" src="https://user-images.githubusercontent.com/56660047/200214233-8a66c99d-21ae-4973-8552-fee1fae97f6e.png">

Reply:
what's the benefit of grouping the event type with the events themselves?

Reply:
what's the benefit of grouping the event type with the events themselves?

Reply:
Its not just the type. The `type` field allow you to include the event specific data, so the Event message can contain `swap`, `deposit`, or `withdraw` event data in addition to the data common to events, and is then identifiable as being from one of those categories. So I can pass a list of `Event` messages to a mapping or store module and handle each depending on the event type like here: https://github.com/messari/substreams/blob/steegecs/develop/uniswap-v2/src/lib.rs#:~:text=pub%20fn-,map_pool_deltas,-(events%3A.

Reply:
Though, if we wanted to use a pattern like this with the `Event` message but more general, we could serialize the event data to a string and deserialize it within a module, but I'd think we'd still want another piece of data to be able to identify the `type` or `name` of the event. .

Reply:
> v2/src/lib.rs#:~:text=%3A%3Amap%5D-,pub%20fn%20map_pool_deltas(events%3A%20dex_amm%3A%3AUsageEvents)%20%2D%3E,%7D,-%23%5Bsubstreams.